src/app_settings.c

```diff
@@ -7,72 +7,37 @@
 #include <zephyr/logging/log.h>
 LOG_MODULE_REGISTER(app_settings, LOG_LEVEL_DBG);
 
-#include <net/golioth/settings.h>
-
+#include <golioth/client.h>
+#include <golioth/settings.h>
 #include "main.h"
 #include "app_settings.h"
 
-static struct golioth_client *client;
-
 static int32_t _loop_delay_s = 60;
+#define LOOP_DELAY_S_MAX 43200
+#define LOOP_DELAY_S_MIN 0
 
 int32_t get_loop_delay_s(void)
 {
 	return _loop_delay_s;
 }
 
-enum golioth_settings_status on_setting(const char *key, const struct golioth_settings_value *value)
-{
-
-	LOG_DBG("Received setting: key = %s, type = %d", key, value->type);
-	if (strcmp(key, "LOOP_DELAY_S") == 0) {
-		/* This setting is expected to be numeric, return an error if it's not */
-		if (value->type != GOLIOTH_SETTINGS_VALUE_TYPE_INT64) {
-			LOG_DBG("Received LOOP_DELAY_S is not an integer type.");
-			return GOLIOTH_SETTINGS_VALUE_FORMAT_NOT_VALID;
-		}
-
-		/* Limit to 12 hour max delay: [1, 43200] */
-		if (value->i64 < 1 || value->i64 > 43200) {
-			LOG_DBG("Received LOOP_DELAY_S setting is outside allowed range.");
-			return GOLIOTH_SETTINGS_VALUE_OUTSIDE_RANGE;
-		}
-
-		/* Only update if value has changed */
-		if (_loop_delay_s == (int32_t)value->i64) {
-			LOG_DBG("Received LOOP_DELAY_S already matches local value.");
-		} else {
-			_loop_delay_s = (int32_t)value->i64;
-			LOG_INF("Set loop delay to %d seconds", _loop_delay_s);
-
-			wake_system_thread();
-		}
-		return GOLIOTH_SETTINGS_SUCCESS;
-	}
-
-	/* If the setting is not recognized, we should return an error */
-	return GOLIOTH_SETTINGS_KEY_NOT_RECOGNIZED;
-}
-
-int app_settings_init(struct golioth_client *state_client)
+static enum golioth_settings_status on_loop_delay_setting(int32_t new_value, void *arg)
 {
-	client = state_client;
-	int err = app_settings_register(client);
-	return err;
-}
-
-int app_settings_observe(void)
-{
-	int err = golioth_settings_observe(client);
-	if (err) {
-		LOG_ERR("Failed to observe settings: %d", err);
-	}
-	return err;
+	_loop_delay_s = new_value;
+	LOG_INF("Set loop delay to %i seconds", new_value);
+	return GOLIOTH_SETTINGS_SUCCESS;
 }
 
-int app_settings_register(struct golioth_client *settings_client)
+int app_settings_register(struct golioth_client *client)
 {
-	int err = golioth_settings_register_callback(settings_client, on_setting);
+	struct golioth_settings *settings = golioth_settings_init(client);
+
+	int err = golioth_settings_register_int_with_range(settings,
```

Comment:
It is currently [expected Golioth Firmware SDK behavior](https://github.com/golioth/firmware-issue-tracker/issues/419) that this registration will result in two notifications that the setting was received.

```
[00:00:07.797,393] <dbg> app_state: async_handler: State successfully set
[00:00:07.815,063] <inf> app_settings: Set loop delay to 10 seconds
[00:00:07.835,510] <inf> app_settings: Set loop delay to 10 seconds
[00:00:07.855,255] <inf> golioth_fw_update: Waiting to receive OTA manifest
```

Reply:
Basically the registration call does an Observe and a Get so the server sends the settings data twice. This will be improved in the SDK but requires some design work first so we will need to be okay with this behavior for now. I left the comment because I spent more than a few minutes wonder if I screwed up the migration :joy: